packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts

```diff
@@ -0,0 +1,317 @@
+  // We do that here as middleware, transforming the server response before passing
+  // it up to VSCode.
+
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const items = (await next(
+      document,
+      position,
+      context,
+      token
+    )) as ProtocolCompletionItem[];
+
+    return expandPlaceholders(await filterByContext(items));
+  }
+};
+
+interface SoqlItemContext {
+  sobjectName: string;
+  fieldName?: string;
+  notNillable?: boolean;
+}
+
+async function filterByContext(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const filteredItems: ProtocolCompletionItem[] = [];
+
+  for (const item of items) {
+    if (
+      !EXPANDABLE_ITEM_PATTERN.test(item.label) &&
+      item?.data?.soqlContext?.sobjectName
+    ) {
+      const objMetadata = await retrieveSObject(
+        item.data.soqlContext.sobjectName
+      );
+      const fieldMeta = objMetadata.fields.find(
+        field => field.name === item.data.soqlContext.fieldName
+      );
+      if (fieldMeta && !operatorsByType[fieldMeta.type].includes(item.label)) {
+        continue;
+      }
+    }
+
+    filteredItems.push(item);
+  }
+
+  return filteredItems;
+}
+
+async function expandPlaceholders(
+  items: ProtocolCompletionItem[]
+): Promise<ProtocolCompletionItem[]> {
+  const expandedItems = [...items];
+
+  for (const [index, item] of items.entries()) {
+    const parsedCommand = item.label.match(EXPANDABLE_ITEM_PATTERN);
+    if (parsedCommand) {
+      const commandName = parsedCommand[1];
+
+      const handler = expandFunctions[commandName];
+      if (!handler) {
+        console.log(`Unknown SOQL completion command ${commandName}!`);
+      }
+
+      expandedItems.splice(
+        index,
+        1,
+        ...(await handler(item?.data?.soqlContext))
+      );
+    }
+  }
+
+  return expandedItems;
+}
+
+const expandFunctions: {
+  [key: string]: (
+    soqlContext?: SoqlItemContext
+  ) => Promise<ProtocolCompletionItem[]>;
+} = {
+  SOBJECTS_PLACEHOLDER: async (
+    soqlContext?: SoqlItemContext
+  ): Promise<ProtocolCompletionItem[]> => {
+    try {
+      const sobjectItems = (await retrieveSObjects()).map(objName => {
+        const item = new ProtocolCompletionItem(objName);
+        item.kind = CompletionItemKind.Class;
+        return item;
+      });
+
+      return sobjectItems;
+    } catch (metadataErrors) {
+      return [];
+    }
+  },
+
+  SOBJECT_FIELDS_PLACEHOLDER: async (
+    soqlContext?: SoqlItemContext
+  ): Promise<ProtocolCompletionItem[]> => {
+    if (!soqlContext?.sobjectName) {
+      console.log('SOBJECT_FIELDS_PLACEHOLDER missing `sobjectName`!');
+      return [];
+    }
+
+    try {
+      const objMetadata = await retrieveSObject(soqlContext.sobjectName);
+      const sobjectFields = objMetadata.fields.reduce((fieldItems, field) => {
+        const icon = iconsByFieldType[field.type];
```

Comment:
Not sure I understand the purpose of these icons. Is it just for decoration or is there a functional purpose for using them ?

Reply:
It shows the user the type of each field.  I personally don't like the icons, but do find the information useful. It could be text instead.

In addition to the icons (which are currently displayed on all items of the list), I've included the textual name of the type to the right of the selected item only. You can see it on the videos above, but here's a screenshot example:
![image](https://user-images.githubusercontent.com/152839/104022187-b8150500-519e-11eb-94f6-f053705378ed.png)

Reply:
I've never seen a language server use emojis and it could be confusing to decode for users (as much as i love using emojis 🎉 ). I think we should just stick with displaying the type via the CompletionItem API like you have already.

Reply:
@dobladez let's get rid of them, as @brpowell pointed out this is not how the rest of the language servers display information and we want to be consistent with the rest since this will eventually be used in Apex code.

Reply:
done